### PR TITLE
Plane/Copter: Allow Landing to terminate the vehicle

### PR DESCRIPTION
**Pull request by WickedShell**
This adds support for using AP_Landing to terminate the vehicle in flight. At the moment the only landing type allowed to terminate the vehicle is deepstall landings. If AP_Landing doesn't handle the termination request then the behavior is unchanged.

This PR is an extension of #6666 

This has been flight tested and works as expected. The deepstall landing also has a couple of small enhancements, primarily the actual landing code is capable of handling the EKF blowing up on it (and then only relying on the gyro to stabilize the aircraft to the ground). 

The main open question on this is that I added a parameter to deepstall to signify if deepstall landings could be used for terminations. I'm wondering if this should just be a common AP_Landing parameter that must be true, then AP_Landing can ask the landing type if it can actually perform the landing. The advantage of this approach is that it avoids more and more parameters as we create more landing types. The downside is someone might thing they could do a glide slope termination. I've flagged this as a dev call topic to get an answer to this question.

### CODE REVIEW

**Comment by WickedShell** [comment]
Rebased and updated to move the landing terminate action into AFS.

I know the code comment is wrong for AFS_TERM_ACTION already but I wanted to discuss that before I fixed it. Particularly I think it would make more sense to split this into each vehicle's afs_*.cpp class as they all terminate differently, and currently copter users of AFS are fed a pile of information about plane termination behaviour. The only hesitation I have is do other GCS's (QGC/MP) resolve the docs from the vehicle first and common second? (IE can I provide a generic term action documentation and then a specific vehicle one in each afs_) or do I need to cut the docs from AFS and let them live in just the vehicle side?

@DonLakeFlyer @meee1 What order does the GCS resolve docs from? Is it the vehicle specific side first, then the libraries, or do you look at libraries first?

**Comment by OXINARF** [comment]
@WickedShell I guess it will depend on the order in which the vehicle lists the paths and also on the parser implementation - I think ground control stations download the files generated by autotest, so you should be able to test it. By the way, we will soon have a way to have different values and descriptions for vehicles.

**Comment by WickedShell** [comment]
@OXINARF Oh? I thought they all tried to use stable ones periodically/match the installed version rather then master. Are you suggesting I test by merging, or is there some other way to test?

**Comment by OXINARF** [comment]
@WickedShell Maybe I'm wrong, but I'm under the impression that they try to get master. Regarding testing, running the param_parse.py script will emit the parameters in multiple formats, you can check one of them to see what happens.

**Comment by WickedShell** [comment]
@OXINARF I'm pretty sure what the param part will look like, I'm just not sure what the GCS will actually do with that (or how to point other GCS's at the generated param set to validate against)

**Comment by OXINARF** [comment]
@WickedShell that will depend on GCSs, I'm not sure there is something you can do about it. Peter's work to support different descriptions and values for vehicles is in my opinion the way forward. I think this was discussed in yesterday's dev call but unfortunately I couldn't be there (and the minutes don't say much).

**Comment by WickedShell** [comment]
@OXINARF the conclusion from the call is Tridge is actually happy with just changing the docs to say go to the wiki for documentation because of how scary/wrong it can go. Long term peters is the best, but that simplifies the docs here. I just need to push that up, but I'm otherwise ready for a review at this point.

**Comment by peterbarker** [comment]
On Wed, 6 Sep 2017, Francisco Ferreira wrote:

> for vehicles is in my opinion the way forward. I think this was discussed in
> yesterday's dev call but unfortunately I couldn't be there (and the minutes
> don't say much).

I think there was a general agreement that my patches there are a good way 
forward, but that a link to the Wiki will do in the short term.

I find it hard to take notes when I'm talking, so the notes on any 
APSync / dataflash stuff are usually crap.

File: ArduCopter/afs_copter.cpp [hunk 1]
```diff
@@ -17,20 +17,26 @@ AP_AdvancedFailsafe_Copter::AP_AdvancedFailsafe_Copter(AP_Mission &_mission, AP_
  */
 void AP_AdvancedFailsafe_Copter::terminate_vehicle(void)
 {
-    // stop motors
-    copter.motors->set_desired_spool_state(AP_Motors::DESIRED_SHUT_DOWN);
-    copter.motors->output();
+    if (_terminate_action == LAND) {
+        if (copter.control_mode != LAND) {
```
Comment (magicrub) [inline ArduCopter/afs_copter.cpp]: this IF is not needed, the set_mode function will just return if the mode matches. simply always call it

File: libraries/AP_AdvancedFailsafe/AP_AdvancedFailsafe.h [hunk 2]
```diff
@@ -45,6 +45,11 @@ class AP_AdvancedFailsafe
         STATE_GPS_LOSS        = 3
     };
 
+    enum terminate_action {
+        TERMINATE = 42,
```
Comment (magicrub) [inline libraries/AP_AdvancedFailsafe/AP_AdvancedFailsafe.h]: "LAND" is a bad stand-alone enum name. Please make this:
```
enum terminate_action {
TERMINATE_ACTION_TERMINATE = 42,
TERMINATE_ACTION_LAND = 43
};

```

**Comment by magicrub** [comment]
looks good except for the copter.set_mode() comment. 

I am a little concerned with how much L1 stuff is in the Landing library. Would be worth moving it to AP_L1 soon before it gets out of hand. We really need to abstract out the wind detection logic to some helper functions because other modules will want to use it. For example: https://github.com/ArduPilot/ardupilot/issues/2222

File: ArduCopter/afs_copter.cpp (hunk 1 above)
Comment (WickedShell) [reply ArduCopter/afs_copter.cpp]: Didn't know what copter did with set_mode, I guess I should have peaked at it to see what happened if you repeatedly called the same mode.

**Comment by WickedShell** [comment]
Addressed Tom's comments.

**Comment by magicrub** [comment]
Thanks, merged!

**Pull request merged**